server/src/main/java/org/elasticsearch/cluster/ClusterModule.java

```diff
@@ -114,6 +115,7 @@ public ClusterModule(Settings settings, ClusterService clusterService, List<Clus
     public static List<Entry> getNamedWriteables() {
         List<Entry> entries = new ArrayList<>();
         // Cluster State
+        registerClusterCustom(entries, RepositoriesState.TYPE, RepositoriesState::new, RepositoriesState::readDiffFrom);
```

Comment:
@ywelsch hmm on second thought:

> This one is a bit odd, as it now has to keep a map for each repo again, and needs clean-up when repos are added / removed.

This seems to be the only real downside though? And it's not really a downside since if I move this stuff into `RepositoryMetaData` I get the some new complication from having to adjust `RepositoriesService` to now use some custom comparisons on `RepositoryMetaData` that ignores the `generation` related fields to see if a repo has changed. 

The other downside is, that we now have the `generation` logic leak into all the `Repository` implementations in a bunch of new spots. `CCR` and all the wrapping repositories don't really care about the generation and it's entirely `BlobStoreRepository` specific. Would we then make adjustments to the serialization of e.g. `GetRepositoriesResponse` so that it doesn't include the generation? We also lose some incrementality/efficiency in the serialization `ClusterState` as a result of mixing somewhat dynamic and pretty static things in it don't we?

I think persisting this state across full-cluster restarts is a good thing still, but I'm not so sure about using `RepositoryMetaData` here having tried implementating this practically now.
Still think it's worth it and make adjustments to things like `GetRepositoriesResponse` accordingly?

Reply:
`RepositoriesService to now use some custom comparisons on RepositoryMetaData that ignores the generation related fields to see if a repo has changed`

I think this is ok (it's similar to how we have it with most items in the cluster state). Further down the line, we should not reinit the repository implementation when there is an update from the CS, but leave it to the implementation to decide whether it needs complete reinitialization or not. This will allow us to do dynamic throttling (e.g. allow user to dynamically change `max_restore_bytes_per_sec` which directly takes place on an ongoing restore). It's a bit bigger investment right now, but will hopefully pay off further down the line.

Regarding serialization, we can make RepositoryMetaData implement Diffable to have it more smartly serialize changes (this would benefit any other field in it as well). I don't expect this to matter much though.

I also think we won't need to adapt `GetRepositoriesResponse` but can just conditionally expose the additional fields by parameterizing toXContent if need be (I'm even fine exposing these fields on the repositories API and not having two variants).